extractClosestPoints() from simplex made more robust to degenerate simplex

SeanCurtis-TRI opened the pull request:
Issue #293 shows a case where the result of the GJK between two separated object is a degenerate simplex. Specifically, a 3-simplex consisting of co-linear vertices. This leads to zero dividing zero which causes NaN to propagate everywhere.

This PR has two curated commits:

1. Added a test to reproduce the scenario described in #293.
2. Correction to the issue. The correction includes:
  - Add degeneracy tests to 3-simplex and 2-simplex extraction algorithms and degradation mechanism (i.e., a degenerate triangle is processed as a line segment and a degenerate line segment is processed as a point).
  - Add unit tests on extractClosestPoints() and all of the supporting methods
  - Updated documentation on the algorithms in place.

Resolves #293 
replaces #295

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/flexible-collision-library/fcl/296)
<!-- Reviewable:end -->

SeanCurtis-TRI commented:
@tongtybj Please take a look at this solution. Sorry for the lag in getting this PR in. Robustly solving the problem took some more rigor.

SeanCurtis-TRI commented:
+@sherm1 

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LDx8Bov8YXNARM4BmGm:bal7hz4)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Hold off on review; I've broken CI.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LDxH5YvBgDCk1MRUWrK:b-gl6jl)*
<!-- Sent from Reviewable.io -->

tongtybj commented:
@SeanCurtis-TRI 

Thank you for your kind commit. I carefully read your improvement.
I also tested with other code, it works well without any errors.

LGTM.

SeanCurtis-TRI commented:
There was a mac CI bug due to the known issue of linking against a single-precision libccd on mac (#291). This most recent modification *should* address that issue.

---

Review status: 2 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LEVG1qe1cXUvNifqfFy:b5nj058)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
@sherm1 CI is all clear. Can you merge for me.

---

Review status: 1 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LEq0yTW-PDvyexudVM7:b-bmrs9r)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
+@hongkai-dai -@sherm1 (per f2f)

---

Review status: 1 of 4 files reviewed, all discussions resolved (waiting on @tongtybj and @sherm1)

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LEq2e759xx22DBVYJI8:b-jwm3fp)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 2 of 3 files at r3.
Review status: 3 of 4 files reviewed, 11 unresolved discussions (waiting on @tongtybj and @SeanCurtis-TRI)

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1017 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEuB-0I9RwWjtVQFc8R:-LEuB-0J5HPQUnoTScXN:b-y1u7q8) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1017)):*
> ```objc
>   //
>   // For dimension i, two values are considered the same if:
>   //   |pᵢ - qⱼ| <= ε·max(1, |pᵢ|, |pⱼ|)
> ```

Should it be `qᵢ` instead of `qⱼ`?

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1026 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEuBPUj-GWVqqnCWCAB:-LEuBPUj-GWVqqnCWCAC:b4mz1qy) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1026)):*
> ```objc
>   // NOTE: Wrapping "1.0" with ccd_real_t accounts for mac problems where ccd
>   // is actually float based.
>   ccd_real_t scales[3] = {
> ```

It might be worthwhile to put the computation of `scales` into a for loop, and do an early exit of the for loop, when the condition |pᵢ - qᵢ| <= ε·max(1, |pᵢ|, |qᵢ|) fails. 
```C++
for (int i = 0; i < 3; ++i) {
  const ccd_real_t scale = max({ccd_real_t{1.0}, abs(p.v[i]), abs(q.v[i])}) * eps
  const ccd_real_t delta = abs(p.v[i] - q.v[i]);
  if (delta > scale) return false;
}
return true;
```

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1047 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEuCV8WASIepZenDaUL:-LEuCV8WASIepZenDaUM:b-xb9ba3) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1047)):*
> ```objc
>                                   const ccd_vec3_t& c) {
>   // First coincidence condition.
>   if (are_coincident(a, b) || are_coincident(a, c)) return true;
> ```

Maybe also check `are_coincident(b, c)` here? If B and C coincide, then it will fail the second test, so it is correct not to check if B and C coincide. But it looks to me checking if B and C coincide is cheaper than checking colinearlity?

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1219 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvvNIo6PrERCCyeusl:-LEvvNIo6PrERCCyeusm:bsqpw5e) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1219)):*
> ```objc
>     //  β = |r_Ap × r_AC| / |r_AB × r_AC|
>     //  β = |r_Ap × r_AC| / |n|                  n ≙ r_AB × r_AC
>     //  β = n̂·(r_Ap × r_AC) / n̂·n                This step arises from the fact
> ```

What is `n̂`? It is better to say that n̂ is the normalized `n` here.

---

*[test/test_fcl_distance.cpp, line 313 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvwh5s-e5WTfqXweYq:-LEvwh5s-e5WTfqXweYr:b-ljhwv4) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/test_fcl_distance.cpp#L313)):*
> ```cpp
> template <typename S>
> void NearestPointFromDegenerateSimplex() {
>   // Tests an historical bug. In certain configurations, the distance query
> ```

s/an/a

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 216 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvyAml12HAPcKa5BvR:-LEvyAml12HAPcKa5BvS:bz44o87) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L216)):*
> ```cpp
>   // The test wants to provide proof as to the boundary of the tolerance
>   // envelope. Therefore, it will construct two triangles. One whose smallest
>   // angle is θₑ - δ and the other triangle's smallest angle is θₑ - δ, for
> ```

One should be θₑ + δ

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 229 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvyQlkBV-fqJsCC_ob:-LEvyQlkBV-fqJsCC_oc:b29ct0m) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L229)):*
> ```cpp
>   //
>   // We'll do it by construction:
>   //   1. Pick a vector perpendicular to [1, 1, 1] (e.g., v = [1, 1, -2]).
> ```

Pick a vector `v`

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 231 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvyywg09--YOOZMyLy:-LEvyywg09--YOOZMyLz:bp93v3w) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L231)):*
> ```cpp
>   //   1. Pick a vector perpendicular to [1, 1, 1] (e.g., v = [1, 1, -2]).
>   //   2. C' = B + v * |B|·tan(θ). This produces a point C' that:
>   //     a) forms a triangle where ∠C`AB < θ, but
> ```

I am not sure if this is a correct statement. Whether ∠C`AB is less than θ o not depends on the magnitude of `v`. If `v` ha unit length, it looks to me ∠C`AB = θ; if `|v| < 1`, then ∠C`AB < θ; otherwise ∠C`AB > θ 

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 254 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvzrk90gBW9BB1-dME:-LEvzrk90gBW9BB1-dMF:brlqzkm) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L254)):*
> ```cpp
>   a << 0, 0, 0;
>   b << 1, 1, 1;
>   Vector3d v = Vector3d(1, 1, -2).normalized();
> ```

const Vector3d

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 305 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEw0Q9m-zREdWLCgKFz:-LEw0Q9m-zREdWLCgKG-:buyaa6h) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L305)):*
> ```cpp
>     const Vector3d& dummy1{-1, -2, -3};
>     const Vector3d& dummy2{-2, -3, -4};
>     ccd_vec3_t p0 = eigen_to_ccd(dummy1);
> ```

Better to rename `p0` with `p1`, and `p1` with `p2`, to match the documentation below.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 558 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEw2kW51iucjoBNoBg2:-LEw2kW51iucjoBNoBg3:b1gce1x) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L558)):*
> ```cpp
> 
>   // Now set up co-linear configuration. Vertex 2 will simply be a linear
>   // combination of vertex 0 and vertex1. v2 = v0 + 2 * v1.
> ```

I do not think this gives a co-linear configuration. For co-linear configuration, it should be v2 = s * v0 + (1-s) * v1, namely the coefficients of v0 and v1 sum up to 1. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed, 11 unresolved discussions (waiting on @SeanCurtis-TRI)

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Thanks @hongkai-dai.  Comments addressed. PTAL

---

Review status: 1 of 4 files reviewed, 11 unresolved discussions (waiting on @hongkai-dai and @SeanCurtis-TRI)

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1017 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEuB-0I9RwWjtVQFc8R:-LEyv78rBsZsLDVh0Qzf:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1017)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Should it be `qᵢ` instead of `qⱼ`?
</blockquote></details>

Done

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1026 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEuBPUj-GWVqqnCWCAB:-LEyvgmU1xd9pnpfnd7U:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1026)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

It might be worthwhile to put the computation of `scales` into a for loop, and do an early exit of the for loop, when the condition |pᵢ - qᵢ| <= ε·max(1, |pᵢ|, |qᵢ|) fails. 
```C++
for (int i = 0; i < 3; ++i) {
  const ccd_real_t scale = max({ccd_real_t{1.0}, abs(p.v[i]), abs(q.v[i])}) * eps
  const ccd_real_t delta = abs(p.v[i] - q.v[i]);
  if (delta > scale) return false;
}
return true;
```
</blockquote></details>

Done

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1047 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEuCV8WASIepZenDaUL:-LEywP7sARMguTnqepD1:b-hrlgxz) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1047)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe also check `are_coincident(b, c)` here? If B and C coincide, then it will fail the second test, so it is correct not to check if B and C coincide. But it looks to me checking if B and C coincide is cheaper than checking colinearlity?
</blockquote></details>

I'm going to make a counter argument. It is cheaper *when* b and c are co-linear. However, if we expect that it's a rare case, we are now paying the extra cost *every time*.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1219 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvvNIo6PrERCCyeusl:-LEyxEog1Cqs-OfEeNj3:bqmjwcz) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1219)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

What is `n̂`? It is better to say that n̂ is the normalized `n` here.
</blockquote></details>

Added.

---

*[test/test_fcl_distance.cpp, line 313 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvwh5s-e5WTfqXweYq:-LEyxJkH-joKn6bSJC98:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/test_fcl_distance.cpp#L313)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

s/an/a
</blockquote></details>

Done

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 216 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvyAml12HAPcKa5BvR:-LEyx_4ZEx5KT6yyv1uT:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L216)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

One should be θₑ + δ
</blockquote></details>

Done

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 229 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvyQlkBV-fqJsCC_ob:-LEyxbbYE47Omr0ankyG:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L229)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Pick a vector `v`
</blockquote></details>

Done

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 231 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvyywg09--YOOZMyLy:-LEyy9w21ISR6CQmmhtD:bqa6w6o) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L231)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I am not sure if this is a correct statement. Whether ∠C`AB is less than θ o not depends on the magnitude of `v`. If `v` ha unit length, it looks to me ∠C`AB = θ; if `|v| < 1`, then ∠C`AB < θ; otherwise ∠C`AB > θ 
</blockquote></details>

I believe making it `v̂`corrects the problem (after all, the *code* was `v̂`.)

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 254 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEvzrk90gBW9BB1-dME:-LEyyJKf4bfbUhT8oFCK:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L254)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

const Vector3d
</blockquote></details>

Done

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 305 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEw0Q9m-zREdWLCgKFz:-LEyyVcX2T-yHKJGvg1r:b-825m3c) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L305)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Better to rename `p0` with `p1`, and `p1` with `p2`, to match the documentation below.
</blockquote></details>

Or change the documentation so I can stay zero-indexed.  :)

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp, line 558 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-LEw2kW51iucjoBNoBg2:-LEyz4tS-oI_7NiFcVNf:b-npm5j5) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e7a9ddea2b07bc8b0dc92996f6ce901dee174cfa/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl_extractClosestPoints.cpp#L558)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I do not think this gives a co-linear configuration. For co-linear configuration, it should be v2 = s * v0 + (1-s) * v1, namely the coefficients of v0 and v1 sum up to 1. 
</blockquote></details>

So, I've updated the intent described in the documentation and corrected the math. PTAL.

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LEzDQ_b0IIMFKic3lwo:b-sfibam)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r2, 3 of 3 files at r4.
Review status: 2 of 4 files reviewed, all discussions resolved (waiting on @hongkai-dai)

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LEzGk2j2QharB3Yps5O:bnfp4nl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
@sherm1 It looks like it's ready for merging. I've double checked the CI failure and it's the known failure.

---

Review status: 2 of 4 files reviewed, all discussions resolved (waiting on @hongkai-dai)

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LFLCvq_9Wul6Ajc8R51:b-fngjcb)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Wow! Looks great. Awesome tests! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r2, 1 of 3 files at r4, 2 of 2 files at r5.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/296#-:-LFLMh9zEMa36mS6hwUP:b-6n2vzf)*
<!-- Sent from Reviewable.io -->
